pallets/pallet-validator/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+			new_node_validator_account: T::AccountId,
+		) -> DispatchResultWithPostInfo {
+			let address = ensure_signed(origin)?;
+
+			let mut validator = <Validator<T>>::get(&address)
+				.ok_or(DispatchError::from(Error::<T>::ValidatorNotFound))?;
+
+			ensure!(
+				validator.state == types::ValidatorRequestState::Validating,
+				Error::<T>::ValidatorNotValidating
+			);
+
+			// Remove the old validator and rotate session
+			substrate_validator_set::Pallet::<T>::remove_validator(
+				frame_system::RawOrigin::Root.into(),
+				validator.validator_node_account.clone(),
+			)?;
+			Self::deposit_event(Event::ValidatorRemoved(validator.clone()));
+
+			// Set the new validator node account on the validator struct
+			validator.validator_node_account = new_node_validator_account.clone();
+			<Validator<T>>::insert(address, &validator);
+
+			// Add the new validator and rotate session
+			substrate_validator_set::Pallet::<T>::add_validator(
+				frame_system::RawOrigin::Root.into(),
+				new_node_validator_account.clone(),
+			)?;
+			Self::deposit_event(Event::ValidatorActivated(validator));
+
+			Ok(().into())
+		}
+
+		/// Bond an account to to a validator account
+		/// Just proves that the stash account is indeed under control of the validator account
+		#[pallet::weight(100_000_000)]
+		pub fn bond(
+			origin: OriginFor<T>,
+			validator: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResultWithPostInfo {
+			let stash = ensure_signed(origin.clone())?;
+
+			if <Bonded<T>>::contains_key(&stash) {
+				Err(Error::<T>::AlreadyBonded)?
+			}
+			let validator = T::Lookup::lookup(validator)?;
+
+			// Cannot bond with yourself
+			if validator == stash {
+				Err(Error::<T>::CannotBondWithSameAccount)?
+			}
+
+			<Bonded<T>>::insert(&stash, &validator);
+
+			Self::deposit_event(Event::Bonded(stash.clone()));
+
+			Ok(().into())
+		}
+
+		/// Approve validator (council)
+		/// Approves a validator to be added as a council member and
+		/// to participate in consensus
+		#[pallet::weight(100_000_000)]
+		pub fn approve_validator(
+			origin: OriginFor<T>,
+			validator_account: T::AccountId,
+		) -> DispatchResultWithPostInfo {
+			T::CouncilOrigin::ensure_origin(origin)?;
+
+			let mut validator = <Validator<T>>::get(&validator_account)
+				.ok_or(DispatchError::from(Error::<T>::ValidatorNotFound))?;
+
+			// Set state to approved
+			validator.state = types::ValidatorRequestState::Approved;
+			<Validator<T>>::insert(validator_account.clone(), &validator);
+
+			// Add the validator as a council member
+			pallet_membership::Module::<T, pallet_membership::Instance1>::add_member(
+				frame_system::RawOrigin::Root.into(),
+				validator_account.clone(),
+			)?;
+
+			Self::deposit_event(Event::ValidatorApproved(validator));
+
+			Ok(().into())
+		}
+
+		/// Remove validator (council)
+		/// Removes a validator from:
+		/// 1. Council
+		/// 2. Storage
+		/// 3. Consensus
+		/// Can only be decided by the council
```

Comment:
wouldn't we allow people to resign from being a validator themselves? Seems better than having them turn off their validator node and wait for the council to notice this and vote for removal.

Reply:
Added an extrinsic for the user to remove his validator